services/cd-service/pkg/repository/transformer.go

```diff
@@ -329,13 +330,66 @@ func (c *CreateApplicationVersion) calculateVersion(bfs billy.Filesystem) (uint6
 				return 0, err
 			}
 		} else {
-			return 0, grpc.AlreadyExistsError(ErrReleaseAlreadyExist)
+			// check if version differs
+			if c.sameAsExisting(state, c.Version) {
+				return 0, grpc.AlreadyExistsSameError(ErrReleaseAlreadyExist)
+			}
+			return 0, grpc.AlreadyExistsDifferentError(ErrReleaseAlreadyExist)
 		}
 		// TODO: check GC here
 		return c.Version, nil
 	}
 }
 
+func (c *CreateApplicationVersion) sameAsExisting(state *State, version uint64) bool {
+	fs := state.Filesystem
+	releaseDir := releasesDirectoryWithVersion(fs, c.Application, version)
+	if c.SourceCommitId != "" {
+		existingSourceCommitId, err := util.ReadFile(fs, fs.Join(releaseDir, "source_commit_id"))
+		if err != nil || string(existingSourceCommitId) != c.SourceCommitId {
+			return false
+		}
+	}
+	if c.SourceAuthor != "" {
+		existingSourceAuthor, err := util.ReadFile(fs, fs.Join(releaseDir, "source_author"))
+		if err != nil || string(existingSourceAuthor) != c.SourceAuthor {
+			return false
+		}
+	}
+	if c.SourceMessage != "" {
+		existingSourceMessage, err := util.ReadFile(fs, fs.Join(releaseDir, "source_message"))
+		if err != nil || string(existingSourceMessage) != c.SourceMessage {
+			return false
+		}
+	}
+	if c.DisplayVersion != "" {
+		existingDisplayVersion, err := util.ReadFile(fs, fs.Join(releaseDir, "display_version"))
+		if err != nil || string(existingDisplayVersion) != c.DisplayVersion {
+			return false
+		}
+	}
+	if c.Team != "" {
+		existingTeam, err := util.ReadFile(fs, fs.Join(releaseDir, "team"))
+		if err != nil || string(existingTeam) != c.Team {
+			return false
+		}
+	}
+	if c.SourceRepoUrl != "" {
+		existingSourceRepoUrl, err := util.ReadFile(fs, fs.Join(releaseDir, "source_repo_url"))
+		if err != nil || string(existingSourceRepoUrl) != c.SourceRepoUrl {
+			return false
+		}
+	}
+	for env, man := range c.Manifests {
+		envDir := fs.Join(releaseDir, "environments", env)
```

Comment:
Should we compare canonical serializations instead?

Reply:
I think we should compare after "normalization by serialization".

But more importantly: if there is a diff, then we should give the user the diff (or at least the line where they diff first), so they can understand the issue - ideally in the http response.

Reply:
done